src/vm/docs/mod.rs

```diff
@@ -50,17 +50,27 @@ const BLOCK_HEIGHT: KeywordAPI = KeywordAPI {
     example: "(> block-height 1000) ;; returns true if the current block-height has passed 1000 blocks."
 };
 
+const CONTRACT_CALLER_KEYWORD: KeywordAPI = KeywordAPI {
+    name: "contract-caller",
```

Comment:
So to be clear, we have three principal-intrinsics:

* `contract-caller` -- the principal of the deepest containing `contract-call!` (or signing tx)
* `tx-sender` -- the principal of the current contract (or signing tx)
* `tx-origin` -- the principal that sent the transaction

Am I understanding this right?

Reply:
No -- there's no `tx-origin` intrinsic.

`contract-caller` is the contract which called the currently executing contract via `contract-call!` or, in the case of a direct execution via transaction, it is equal to `tx-sender`.
`tx-sender` is the principal that sent the transaction, unless it has been changed using `as-contract`

Reply:
So `contract-caller` doesn't change within `as-contract`, but `tx-sender` does?  I think this name-choice is confusing if so -- seeing `contract-caller` for the first time, my first impression is that it was the principal doing the call into the current function (i.e. like `msg.sender`), whereas my first impression of `tx-sender` was that it referred to the principal that sent the actual transaction.  Just my 2 cents.

Reply:
Both change with `as-contract` -- which maybe is the confusingly named thing here, since your intuition is basically correct --

In the absence of `as-contract`, `contract-caller` behaves like `msg.sender` in Eth --- basically it's the calling contract if `contract-call!` is made and otherwise equals `tx-sender`, and `tx-sender` is the actual sender of the transaction.

`as-contract` is provided to allow contracts to _completely_ emulate sending a transaction as if they were any other kind of principal --- it evaluates its argument with `tx-sender` and `contract-caller` set to the current contract (this way, it looks like the contract principal sent a "direct" transaction).

Reply:
Gotcha.  But if I call function B in contract C_B from function A in contract C_A, and call function A with `as-contract`, then within A, `tx-sender` and `contract-caller` are both C_A, but within B, `contract-caller` is C_A but `tx-sender` is C_B?

Reply:
No, `contract-caller` would be `C_A`, since that's the most recent caller of `contract-call!` --

`contract-caller` is the "head" of the `contract-call!` stack.

```
;; CONTRACT C_B
(define-public (foo)
   (begin (print tx-sender)
              (print contract-caller)
              (ok 1)))

;; CONTRACT C_A
(define-public (foo)
   (begin (print tx-sender)
              (print contract-caller)
              ;; while invoking contract-call, the contract-caller is set to C_A (the current contract)
              (contract-call! c_b foo) 
              (as-contract 
                 ;; tx-sender and contract-caller both updated to C_A in the
                 ;;     evaluation of as-contracts arg.
                 (begin (print tx-sender) (print contract-caller)
                 ;; while invoking contract-call, the contract-caller is set to C_A (the current contract)
                 ;;   but, in this context it was _already_ C_A.
                            (contract-call! c_b foo)))))
```

```
$ blockstack-core local launch c_a contract_a.clar /tmp/db 
Contract initialized!
$ blockstack-core local launch c_b contract_b.clar /tmp/db 
Contract initialized!
$ blockstack-core local execute /tmp/db c_b foo SP1CMKS1EMEG18AYXEFJN8GECXF1DDY6CM0P84BQ0
Principal(StandardPrincipal(22, [89, 73, 228, 46, 163, 160, 20, 43, 221, 115, 229, 84, 65, 204, 235, 194, 214, 248, 204, 160]))
Principal(StandardPrincipal(22, [89, 73, 228, 46, 163, 160, 20, 43, 221, 115, 229, 84, 65, 204, 235, 194, 214, 248, 204, 160]))
Principal(StandardPrincipal(22, [89, 73, 228, 46, 163, 160, 20, 43, 221, 115, 229, 84, 65, 204, 235, 194, 214, 248, 204, 160]))
Principal(ContractPrincipal("c_a"))
Principal(ContractPrincipal("c_a"))
Principal(ContractPrincipal("c_a"))
Principal(ContractPrincipal("c_a"))
Principal(ContractPrincipal("c_a"))
Transaction executed and committed. Returned: 1
```

Which is:

```
Transaction Sender
Transaction Sender
Transaction Sender
C_A
C_A
C_A
C_A
C_A
```

Reply:
Ah -- no, I wasn't able to do that -- that should actually be `c_b` launching before `c_a`. I'd originally named them the opposite, and then went through and renamed them, and forgot to swap the names in that command line comment.